File: polyglot/runtime_template_resolver/mjs/src/logger.ts
/**
 * Polyglot Logger Implementation
 */

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

export interface LoggerInterface {
    debug(message: string, context?: Record<string, unknown>): void;
    info(message: string, context?: Record<string, unknown>): void;
    warn(message: string, context?: Record<string, unknown>): void;
    error(message: string, context?: Record<string, unknown>): void;
}

const LOG_LEVELS: Record<LogLevel, number> = {
    debug: 0,
    info: 1,
    warn: 2,
    error: 3,
};

class ConsoleLogger implements LoggerInterface {
    private packageName: string;
    private filename: string;
    private level: number;

    constructor(packageName: string, filename: string, level: LogLevel = 'debug') {
        this.packageName = packageName;
        this.filename = filename;
        this.level = LOG_LEVELS[level];
    }

    private shouldLog(level: LogLevel): boolean {
        return LOG_LEVELS[level] >= this.level;
    }

    private format(level: LogLevel, message: string, context?: Record<string, unknown>): void {
        if (!this.shouldLog(level)) return;

        // In node, filename might be full path, we want basename
        const plainFilename = this.filename.split(/[/\\]/).pop() || this.filename;
        const timestamp = new Date().toISOString();

        // Structured log format
        const output = {
            timestamp,
            level,
            package: this.packageName,
            file: plainFilename,
            message,
            ...context
        };

        console.log(JSON.stringify(output));
    }

    debug(message: string, context?: Record<string, unknown>): void {
        this.format('debug', message, context);
    }

    info(message: string, context?: Record<string, unknown>): void {
        this.format('info', message, context);
    }

    warn(message: string, context?: Record<string, unknown>): void {
        this.format('warn', message, context);
    }

    error(message: string, context?: Record<string, unknown>): void {
        this.format('error', message, context);
    }
}

export const logger = {
    create: (packageName: string, filename: string): LoggerInterface => {
        const envLevel = (process.env.LOG_LEVEL as LogLevel) || 'debug';
        return new ConsoleLogger(packageName, filename, envLevel);
    }
};

File: polyglot/runtime_template_resolver/py/src/runtime_template_resolver/logger.py
import os
import json
import sys
import datetime
from typing import Any, Dict, Protocol, Literal

class LoggerProtocol(Protocol):
    def debug(self, message: str, **kwargs: Any) -> None: ...
    def info(self, message: str, **kwargs: Any) -> None: ...
    def warning(self, message: str, **kwargs: Any) -> None: ...
    def error(self, message: str, **kwargs: Any) -> None: ...

LogLevel = Literal["debug", "info", "warn", "error"]

LOG_LEVELS: Dict[LogLevel, int] = {
    "debug": 0,
    "info": 1,
    "warn": 2,
    "error": 3
}

class ConsoleLogger:
    def __init__(self, package_name: str, filename: str, level: LogLevel = "debug"):
        self.package_name = package_name
        self.filename = os.path.basename(filename)
        self.level_name = level
        self.level_value = LOG_LEVELS.get(level, 0)

    def _should_log(self, level: LogLevel) -> bool:
        return LOG_LEVELS.get(level, 0) >= self.level_value

    def _format(self, level: LogLevel, message: str, **kwargs: Any) -> None:
        if not self._should_log(level):
            return

        output = {
            "timestamp": datetime.datetime.now().isoformat(),
            "level": level,
            "package": self.package_name,
            "file": self.filename,
            "message": message,
            **kwargs
        }
        print(json.dumps(output), file=sys.stdout)

    def debug(self, message: str, **kwargs: Any) -> None:
        self._format("debug", message, **kwargs)

    def info(self, message: str, **kwargs: Any) -> None:
        self._format("info", message, **kwargs)

    def warning(self, message: str, **kwargs: Any) -> None:
        self._format("warn", message, **kwargs)

    def error(self, message: str, **kwargs: Any) -> None:
        self._format("error", message, **kwargs)

class LoggerFactory:
    @staticmethod
    def create(package_name: str, filename: str) -> LoggerProtocol:
        env_level = os.getenv("LOG_LEVEL", "debug").lower()
        # map warning to warn if needed, basic check
        if env_level == "warning": env_level = "warn"
        return ConsoleLogger(package_name, filename, level=env_level) # type: ignore

logger = LoggerFactory()
